Display the nickname or full name based on settings when drafting posts

kondo97 opened the pull request:
#### Summary
A user can configure the Teammate Name Display in the System Console. Since usernames must be unique and cannot contain restricted characters, my organization configures using nicknames or full names.

However, when drafting posts, only the username is displayed in mentions, even when the system is configured to use nicknames or full names. This is not user-friendly.

By aligning the display with the mention list format, improves the clarity of username display. 

The username is still stored in the database. While the nickname or full name is displayed in the draft, it is removed before submission and sent to the server. Therefore, the impact of this implementation is minimized.

#### Ticket Link
Fixes https://github.com/mattermost/mattermost/issues/30038

#### Screenshots
**Before**
Only displaying username is very difficult to understand.

https://github.com/user-attachments/assets/8be81dd6-b1b7-417a-a71c-0561367ec6b2

**After**

https://github.com/user-attachments/assets/4d4f1a08-439f-49d1-aad4-9c1e2a6a6e3c

Here is a preview behavior.

https://github.com/user-attachments/assets/8a21c3fa-002e-44b6-9bad-d872506ea3e6

#### Release Note
```release-note
Display the nickname or full name based on settings when drafting posts.
```

mattermost-build commented:
Hello @kondo97,

Thanks for your pull request! A Core Committer will review your pull request soon. For code contributions, you can learn more about the review process [here](https://developers.mattermost.com/contribute/getting-started/code-review/).

kondo97 commented:
I'm unsure whether the community will accept this change. However, the current specification is confusing and not user-friendly. I hope we can have a good discussion about it.

larkox commented on webapp/channels/src/components/suggestion/at_mention_provider/at_mention_provider.tsx:
```diff
@@ -362,12 +364,34 @@ export default class AtMentionProvider extends Provider {
+                const state = store.getState();
+                const teammateNameDisplay = getTeammateNameDisplaySetting(state)
+
+                if (teammateNameDisplay == 'nickname_full_name') {
+                    const fullname = getFullName(item);
+                    const displayName = item.nickname || fullname;
+                    if (displayName) {
+                        return `@${item.username}(${displayName})`;
+                    }
+                }
+
+                if (teammateNameDisplay == 'full_name') {
+                    const fullname = getFullName(item);
+                    if (fullname) {
+                        return `@${item.username}(${fullname})`;
+                    }
+                }
```
Why not using `displayUsername`? It could be something like...

```
const displayName = displayUsername(user, settings);
if (item.username !== displayName) {
    return `@${item.username}(${displayName})`;
}
```

kondo97 replied:
@larkox 
You're right. I've corrected it.
https://github.com/mattermost/mattermost/pull/30145/commits/d5a64bf73ea07e446636628b9642e9758d1eb3c0

larkox commented on webapp/channels/src/components/advanced_text_editor/use_submit.tsx:
```diff
@@ -373,6 +379,18 @@ const useSubmit = (
             }
         }
 
+        submittingDraft.message = getUsernameMentions(submittingDraft.message).reduce((msg, mention) => {
+            const username = mention.substring(1);
+            const mentionedUser = usersByUsername[username];
+        
+            if (!mentionedUser) {
+                return msg;
+            }
+        
+            const userDisplayName = displayUsername(mentionedUser, teammateNameDisplay);
+            return msg.replace(`@${username}(${userDisplayName})`, `@${username}`);
+        }, submittingDraft.message);
```
Can you add unit tests for this behavior? I would love to see some tests, specially to understand what will happen with corner cases.

kondo97 replied:
I've added some test cases for this logic. Is this sufficient? 
https://github.com/mattermost/mattermost/pull/30145/commits/6d3379e93d377ac0eca016cc67a9a679dd5c9359

larkox reviewed (changes requested):
Thank you for the contribution! It looks quite nice, and it is a quite smart approach. But I have some reservations. I would love to have @hmhealey and @abhijit-singh  feedback on this one.

kondo97 commented:
@larkox 
Thank you for your review!
I have implemented additional changes. Could you review it again?
I forgot to handle the drafting preview, so I have also implemented that. I've added a video in the description.

hmhealey commented:
Thanks for the PR, @kondo97. I like the idea as well, but unfortunately, I don't think this approach is going to be something we can accept. Storing this information in the post draft as text kind of works, but I'd be very worried about whatever bugs and edge cases might come up. There's so much potential for things to go wrong when replacing the text in drafts, and the post textbox is the most important part of the app to have working right, so we have to be very careful when making changes to it.

A couple of the problems I can see with this is that we may not properly remove the extra text in certain cases (like if the draft is sent after a refresh and the user isn't loaded any more) and that the extra text is editable which will likely confuse users who are used to the current behaviour.

The proper way to do this would be to rewrite how we store draft messages entirely so that it uses a structure which could hold more information about things like mentions and be shown to the end user more like a WYSIWYG editor, but that's a much larger project than we're able to work on or help someone else work on right now.

I definitely appreciate the work that you've put in on this, but this isn't likely to be a feature that we're able to accept work on any time soon

kondo97 commented:
@hmhealey 
I predicted that this change would be difficult to accept. I understand that this change carries some risks.
Thank you for your cooperation.
